### ISSUE TITLE
.each() synchronization question

### ISSUE THREAD

**Issue by ericlu88** [petkaantonov/bluebird#513]
We encountered the following synchronization problem while using `.each()`. We couldn't figure out if it is intentionally design like this or a bug based on the documentation, so would like to get any insight from you guys:

```javascript
// Expect: 1 2 3 4 5
// Actual: 1 3 4 5 2
console.log(1);
Promise.resolve([3,4,5])
    .each(function(v) {
        console.log(v);
    });
console.log(2);
```
It seems like the function get executed synchronously. The same problem happens to `.filter()` function too. In the contrary , the following snippet get executed expectedly:
```javascript
// Expect: 1 2 [3 4 5]
// Actual: 1 2 [3 4 5]
console.log(1);
Promise.resolve([3,4,5])
    .then(function(v) {
        console.log(v);
    });
console.log(2);
```

**Comment by petkaantonov** [comment]
It's definitely a bug, the reduce implementation in 2.x is notoriously complicated and each delegates to it as well hence it is affected. The implementation has been redone in 3.x and it shouldn't have this bug. If this is critical for you I can make a patch for 2.9.13.

**Comment by ericlu88** [comment]
Thanks for getting back super fast. Right now we are using a work around by chaining a `.then()` as a pass through to make it really async. It would be awesome if we can have a patch release soon. Also, when are you planning for the 3.x release?

**Comment by petkaantonov** [comment]
Fix is in 2.9.13 release just published to npm. 3.x will be released probably 1-3 weeks from now.

**Comment by ericlu88** [comment]
Cool, just verified the issue get fixed. Thanks a lot!
